File: day_07/day7.jl
#=
day7:
- Julia version: 
- Author: mike.downey
- Date: 2022-12-07

This is based on the first attempt in python, not the modified more object-oriented
version given here.
=#
 
data = readlines(open("day_07/data.txt"))

# Nodes holding the files and directories
mutable struct FileTree
    name::String
    size::Int
    children::Dict
    parent::FileTree
    path::String

    function FileTree(path::Array{String}, name::String, size::Int=0)
        this = new()
        this.children = Dict()
        this.name = name
        this.size = size
        this.parent = this
        this.path = string(join("/",path), name)
        return this
    end
end

function addnode(this::FileTree, node::FileTree)
    this.children[node.name] = node
    node.parent = this
end

getchild(this::FileTree, name::String) = this.children[name]

path = [""]
root = FileTree(path, "/")
cnode = root

function cd(p::String)
    global path, cnode
    if p == "/"
        path = ["/"]
        cnode = root
        return
    end
    if p == ".."
        pop!(path)
        cnode = cnode.parent
        return
    end
    push!(path, p)
    cnode = getchild(cnode, p)
end

for line in data
    bits = split(line," ")
    if bits[1] == "\$"
        if bits[2] == "cd"
            cd(string(bits[3]))
        end
        continue
    end
    # within the output of 'ls'
    fileinfo = parse(Int, replace(string(bits[1]), "dir" => "0"))
    name = string(bits[2])
    newnode = FileTree(path, name, fileinfo)
    addnode(cnode, newnode)
end

sizes = Dict()

function getsize(node::FileTree)
    global sizes
    if node.size != 0
        return node.size
    end
    childsizes = [getsize(n) for n in values(node.children)]
    s = sum(childsizes)
    sizes[node.path] = s
    return s
end

rootsize = getsize(root)
println(rootsize)

# Part 1
println("Part 1")
# Dicts iterate Pairs which can be accessed using .first and .second
total = sum([s.second for s in sizes if s.second<100000])
println(total)

# Part 2
println("Part 2")
unused = 70000000 - sizes["//"]
needtofree = 30000000 - unused
# println(needtofree)
dirs = [s for s in sizes if s.second > needtofree]
second(dir::Pair) = dir.second
sorted = sort(dirs, by=second)
println(sorted[1].second)

File: day_07/day7.py
from aoc.filetree import FileTree
from aoc import aoc


class Disk:
    def __init__(self, script):
        self.path = []
        self.root = FileTree(self.path, "")
        self.cnode = self.root
        self.sizes = {}
        self.parse(script)

    def cd(self, p):
        if p == "/":
            self.path = [""]
            self.cnode = self.root
            return
        if p == "..":
            self.path.pop()
            self.cnode = self.cnode.parent
            return
        self.path.append(p)
        self.cnode = self.cnode.getchild(p)
        return

    def parse(self, script):
        for line in script:
            bits = line.split(" ")
            if bits[0] == "$":
                if bits[1] == "cd":
                    self.cd(bits[2])
                    continue
                if bits[1] == "ls":
                    # read until the next command
                    continue
            sizeordir = int(bits[0]) if bits[0] != "dir" else 0
            newnode = FileTree(self.path, bits[1], sizeordir)
            self.cnode.addnode(newnode)

    def getsize(self, node):
        if node.size != 0:
            return node.size
        s = sum([self.getsize(n) for n in node.children.values()])
        self.sizes[node.path] = s
        return s

    def countem(self):
        return sum([self.sizes[d] for d in self.sizes if self.sizes[d] <= 100000])


def part1(disk):
    disk.getsize(disk.root)
    total = disk.countem()
    print(total)
    assert (total == 1432936 or total == 95437)
    return total


def part2(disk):
    unused = 70000000 - disk.sizes["/"]
    needtofree = 30000000 - unused
    availabledirs = list(filter(lambda x: x >= needtofree, disk.sizes.values()))
    availabledirs.sort()
    dirsize = availabledirs[0]
    assert (dirsize == 272298 or dirsize == 24933642)
    return dirsize


def run(file):
    data = aoc.loadlines(file)
    disk = Disk(data)

    print("Part 1")
    print(part1(disk))
    print("Part 2")
    print(part2(disk))


if __name__ == "__main__":
    print("Test Data")
    run("test.txt")
    print("Actual Data")
    run("data.txt")
